presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionarySelectiveStreamReader.java

```diff
@@ -275,15 +275,12 @@ private void openRowGroup()
     {
         // read the dictionary
         if (!dictionaryOpen && dictionarySize > 0) {
-            if (dictionary.length < dictionarySize) {
-                dictionary = new long[dictionarySize];
-            }
-
-            LongInputStream dictionaryStream = dictionaryDataStreamSource.openStream();
-            if (dictionaryStream == null) {
-                throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Dictionary is not empty but data stream is not present");
-            }
-            dictionaryStream.nextLongVector(dictionarySize, dictionary);
+            LongDictionaryProvider.GetDictionaryResult getDictionaryResult = dictionaryProvider.getDictionary(streamDescriptor, dictionary, dictionarySize);
+            dictionary = getDictionaryResult.dictionaryBuffer;
+            // If using shared dictionary buffer, dictionaryMemoryUsage remains 0. Since flat map
+            // stream readers get reset in the next stripe, we don't need to distinguish carried
+            // over buffers and shared buffers.
+            dictionaryMemoryUsage = getDictionaryResult.reused ? dictionaryMemoryUsage : sizeOf(dictionary);
```

Comment:
This does not seem right, dictionaryMemoryUsage should be set to 0 in close or reset methods. or this statement needs to be . Or what is the problem with this following statement which seems simpler

dictionaryMemoryUsage = getDictionaryResult.reused ? 0 : sizeOf(dictionary);

Reply:
What I'm considering are the following cases:
* for non-flatmap columns, dictionary can be carried over from previous stripe. Dictionary memory usage doesn't change when buffer is reused. The value starts from 0 but never gets reset to 0. close() is never explicitly called in the project, and I assume it's meant for cleanup in dtor.
* for flatmap columns, their column reader only live for one stripe, there's ever only change from 0->nonzero when the buffer is not reused. (maybe the better phrasing would be whether a new buffer is allocated) For flatmap shared dictionary case (in the next PR), we return reused = false when we have to create the first dictionary entry, and true otherwise.